### PR TITLE
Lintang/noncausal attention

### PR DESCRIPTION
**Pull request by lintangsutawika**
# What does this PR do?

I made a few changes I wanted to propose for the noncausal attention for BLOOM.

The idea:
1. `prefix_mask` changed to `prefix_length` as the mask can be easily calculated in the model.
2. `causal_mask` is updated with `prefix_mask` to create a non causal mask which is done along with `causal_mask` in `BloomScaledSoftmax`.

I checked the attention with this
```
from transformers import BloomForCausalLM, BloomForPrefixLM
from transformers import AutoTokenizer

import torch

model_path = "bigscience/bloom-350m"
bloom = BloomForPrefixLM.from_pretrained(model_path)
tokenizer = AutoTokenizer.from_pretrained(model_path)

input_string = "This is the first part."
target_string = "This is the second part."
tokenized_inputs = tokenizer(
    input_string,
    return_tensors="pt"
    )

tokenized_targets = tokenizer(
    target_string,
    return_tensors="pt"
    )

model_inputs = {}
for key in tokenized_inputs.keys():
    model_inputs[key] = torch.concat(
        (tokenized_inputs[key], tokenized_targets[key]),
        -1)

model_inputs["prefix_length"] = tokenized_inputs["input_ids"].shape[1]
output = bloom(**model_inputs, output_attentions=True)
output.attentions[0][0][0]
```

which outputs a causal mask but with a full attention at the left part of the mask.
```
[2.3241e-03, 1.8340e-02, 3.9363e-02, 9.4744e-02, 1.7149e-01, 6.7373e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.6395e-02, 6.9117e-03, 8.4897e-02, 9.7560e-02, 1.4556e-01, 6.4868e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[9.5913e-03, 4.5855e-02, 1.5791e-02, 1.5099e-01, 1.8595e-01, 5.9183e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.2768e-02, 2.6686e-02, 4.7857e-02, 2.0128e-02, 1.0224e-01, 7.9032e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.9771e-02, 3.7443e-02, 6.2693e-02, 1.0858e-01, 3.8452e-02, 7.3306e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.2164e-02, 9.8382e-02, 9.2077e-02, 1.5037e-01, 5.1458e-01, 1.3243e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[2.0006e-03, 1.5787e-02, 3.3883e-02, 8.1554e-02, 1.4762e-01, 5.7994e-01, 1.3922e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[6.2529e-03, 2.6361e-03, 3.2380e-02, 3.7209e-02, 5.5516e-02, 2.4741e-01, 4.3515e-01, 1.8345e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.6100e-03, 7.6970e-03, 2.6506e-03, 2.5344e-02, 3.1213e-02, 9.9342e-02, 1.1204e-01, 5.3565e-01, 1.8446e-01, 0.0000e+00, 0.0000e+00, 0.0000e+00],
[1.2422e-03, 4.5851e-03, 4.6159e-03, 6.8153e-03, 1.7759e-02, 8.4931e-02, 8.6448e-02, 3.1908e-01, 3.2123e-01, 1.5329e-01, 0.0000e+00, 0.0000e+00],
[8.2140e-04, 1.5555e-03, 2.6046e-03, 4.5111e-03, 1.5975e-03, 3.0455e-02, 5.7163e-02, 1.0825e-01, 1.8126e-01, 5.0061e-01, 1.1117e-01, 0.0000e+00],
[1.5725e-04, 1.2719e-03, 1.1903e-03, 1.9439e-03, 6.6524e-03, 1.7120e-03, 1.0943e-02, 8.8510e-02, 8.2837e-02, 2.2270e-01, 4.6295e-01, 1.1914e-01]],
```

Tagging: @haileyschoelkopf

### CODE REVIEW

**Comment by lintangsutawika** [comment]
Added tests:
1. `test_noncausal_attention_bloom.py` -> test if the non-causal attention is correct
2. `test_bloom_generation.py` -> test if it works with generation.

File: src/transformers/models/bloom/modeling_bloom.py [hunk 1]
```diff
@@ -318,7 +301,19 @@ def forward(self, input, mask, max_positions, prefix_mask):
...
+                (prefix_length.shape[-1], 1, max_positions, max_positions),
+                dtype=torch.bool
+            )
+            for idx in range(prefix_length.shape[-1]):
+                prefix_mask_length = int(prefix_length[idx])
+                prefix_mask[idx, :, :prefix_mask_length, :prefix_mask_length] = 1
+            prefix_mask = prefix_mask.to(input.device)
+            causal_mask = (causal_mask+prefix_mask).bool()
+
+        mask_output, padded_causal_mask = self.mask_func(input, mask, causal_mask)
```
Comment (haileyschoelkopf) [inline src/transformers/models/bloom/modeling_bloom.py:316]: I'll need to check this, but why is it `prefix_length.shape[-1]` instead of `prefix_length.shape[0]`, shouldn't the batch dimension be the first dimension?

File: src/transformers/models/bloom/modeling_bloom.py [hunk 2]
```diff
@@ -1036,19 +1031,19 @@ def prepare_inputs_for_generation(self, input_ids, past=None, **kwargs):
 
         # generate a prefix-LM mask, if none is provided.
         # store this in the model class so that it can be reused for a whole generation call. Will be erased on next `generate()` call.
-        if prefix_mask is None and past is None:
-            prefix_mask = torch.ones(input_ids.shape[0], 1, input_ids.shape[1], input_ids.shape[1], dtype=torch.bool, device=input_ids.device)
-            self.prefix_mask = prefix_mask
-        else:
-            prefix_mask = self.prefix_mask
+        if prefix_length is None:
```
Comment (haileyschoelkopf) [inline src/transformers/models/bloom/modeling_bloom.py:1034]: TODO: make this support batched generation, I could do this

**Comment by haileyschoelkopf** [comment]
I can merge these changes and write the tests as unit tests in Transformers. I'll do this today or tomorrow

File: src/transformers/models/bloom/modeling_bloom.py (hunk 2 above)
Comment (lintangsutawika) [reply src/transformers/models/bloom/modeling_bloom.py:1034]: I've actually made wrote this so that it can do batched generation. It's in the `test_generation_bloom.py`

File: src/transformers/models/bloom/modeling_bloom.py (hunk 1 above)
Comment (lintangsutawika) [reply src/transformers/models/bloom/modeling_bloom.py:316]: You're right. Let me change this. This should be in fact `prefix_length.shape[0]`.

File: src/transformers/models/bloom/modeling_bloom.py (hunk 2 above)
Comment (haileyschoelkopf) [reply src/transformers/models/bloom/modeling_bloom.py:1034]: Awesome!!

**Comment by haileyschoelkopf** [comment]
Thanks Lintang! I'll merge this into my PR once you're done so I can work on a few changes to the tests there

**Comment by lintangsutawika** [comment]
Done!
